# Вивчення законів збереження енергії та імпульсу для зіткнення куль
### Лабораторна робота № 1

Виконав студент групи ФІ-81 *Шкаліков Олег Володимирович*  
Перевірив *Димитрієва Наталія Федорівна*

**Мета**: на прикладі зіткнення куль перевірити закони збереження  
**Завдання**:
- виміряти кути відхилення куль
- розрахувати коефіцієнт розсіювання енергії
- розрахувати співвідношення мас куль

## Теоретична частина

![](Lab2.png)  
Підвішена на нитці куля, масою $m_1$, відвена на кут $\alpha_1$, падає на кулю масою $m_2$, яка знаходиться у положенні рівноваги. Внаслідок закону збереження енергії друга куля відхиляється на кут $\alpha_2$. Причому частина енергії розсіюється. В даній роботі вимірюється коефіцієнт розсіяння енергії, який дорівняє:
$$\beta = \frac{E_{початкова} - E_{кінцева}}{E_{початкова}} \Rightarrow E_{кінцева} = (1-\beta)E_{початкова}$$
Крайні точки траекторії розташовані на певнії висоті($h_1$ та $h_2$ відповідно) від рівня рівноваги. У них швидкість дорівнює нулю, тому закон збереження енергії приймає натупний вид:
$$m_2gh_2 = (1-\beta)m_1gh_1$$
Використовуючи тригонометричні співвідношення виразимо висоти через довжину нитки.
$$h_1 = l(1 - \cos{\alpha_1}), h_2 = l(1 - \cos{\alpha_2})$$
Отже,
$$m_2gl(1 - \cos{\alpha_2}) = (1-\beta)m_1gl(1 - \cos{\alpha_1})$$
Якщо взяти однаккові кулі, тобто $m = m_1 = m_2$ отримаємо вираз, що пов'язує кути відхилення з коефіцієнт розсіяння енергії.
$$mgl(1 - \cos{\alpha_2}) = (1-\beta)mgl(1 - \cos{\alpha_1})$$
$$1-\beta = \frac{1 - \cos{\alpha_2}}{1 - \cos{\alpha_1}} \Rightarrow \beta = 1 - \frac{1 - \cos{\alpha_2}}{1 - \cos{\alpha_1}}$$
Тому для обрахунку значення коефіцієнта розсіяння енергії візьмемо дві кулі однакової маси та будемо вимірювати кути відхилення.  

Застосуємо знайдений $\beta$ для обчислення відношення масс будь-яких куль, що мають виміряні кути відхилення. Не важко переконатися, що з закону збереження енергії випливає наступне співвідношення:
$$m_2gl(1 - \cos{\alpha_2}) = (1-\beta)m_1gl(1 - \cos{\alpha_1}) \Rightarrow \frac{m_1}{m_2} = \frac{1 - \cos{\alpha_2}}{(1-\beta)(1 - \cos{\alpha_1})}$$

## Практична частина

Для обрахунку, перетворення та візуалізації результатів експерименту будемо використовувати бібліотеки Pandas та Numpy

In [196]:
import pandas as pd
import numpy as np

Вивантажимо дані експерименту, збережені у csv файлі, та поділимо на чотири набори даних за типом:
1. "велика" куля падає на "велику"
2. "маленька" куля падає на "маленька"
3. "маленька" куля падає на "велику"
4. "велика" куля падає на "маленьку"  
Усього було зроблено по 3 заміри з різним початковим кутом відзилення $\alpha_1$ для кожного типу експерименту.  
Також, для підвишення точності експерименту, дані для кожного кута відхилення $\alpha_1$ було виміряно 5 разів. Для отримання значення для конкретного кута з цих даних візьмемо середнє-арифметичне отриманих кутів. Загалом було отримано 60 значень.

In [136]:
df = pd.read_csv("Lab2.csv")
df['α1'] = df.apply(lambda row: np.radians(row['α1']), axis=1)
df['α2'] = df.apply(lambda row: np.radians(row['α2']), axis=1)
df.columns = ['a1','a2','type']
BB = df[df['type'] == 1][['a1','a2']]
BB = BB.groupby(['a1'], as_index = False).mean()
SS = df[df['type'] == 2][['a1','a2']]
SS = SS.groupby(['a1'], as_index = False).mean()
SB = df[df['type'] == 3][['a1','a2']]
SB = SB.groupby(['a1'], as_index = False).mean()
BS = df[df['type'] == 4][['a1','a2']]
BS = BS.groupby(['a1'], as_index = False).mean()

In [112]:
def calculateBeta(alpha1, alpha2):
    ones = np.ones(len(alpha1))
    a = ones-np.cos(alpha2)
    b = ones-np.cos(alpha1)
    return ones - np.divide(a,b)

Обчислимо коефіцієнта розсіяння енергії для експериментів з однаковими кулями:
1. "велика" куля падає на "велику"
2. "маленька" куля падає на "маленька"

In [115]:
betaBB = calculateBeta(BB['a1'], BB['a2']);
betaSS = calculateBeta(SS['a1'], SS['a2']);
print(betaBB.mean())
print(betaSS.mean())

0.15278523093584798
0.25809575928391765


Ми отримали суттєво різні значення:
1. "велика" куля падає на "велику": $\beta \approx 0.15$
2. "маленька" куля падає на "маленька": $\beta \approx 0.26$

тому надалі, будемо проводити по два обрахунки для кожного зі значень $\beta$

Обчислимо відношення мас для типів експерименту:
3. "маленька" куля падає на "велику"
4. "велика" куля падає на "маленьку"

In [79]:
def calculateMassRelation(alpha1, alpha2, beta):
    ones = np.ones(len(alpha1))
    a = ones - np.cos(alpha2)
    b = (1-beta)*(ones - np.cos(alpha1))    
    return np.divide(a,b)

Для значення $\beta \approx 0.15$

In [207]:
BSMassBBRel = calculateMassRelation(BS['a1'], BS['a2'], betaBB)
SBMassBBRel = calculateMassRelation(SB['a1'], SB['a2'], betaBB)
MassBBRel = (BSMassBBRel.mean() + 1 / SBMassBBRel.mean()) / 2 

print(MassBBRel)
print(BSMassBBRel.mean())
print(SBMassBBRel.mean())

1.5587535197769675
1.3194390093656352
0.5561524832268313


Було обраховано результати для усіх початкових кутів відхилення та взято середньо-арифметичне по кожному типу експерименту:
- "маленька" куля падає на "велику": $\frac{m_1}{m_2} \approx 0.56$
- "велика" куля падає на "маленьку": $\frac{m_1}{m_2} \approx 1.32$

Середнім арифметичним відношення масс "великої" та  "маленької" кулі є: $\approx 1.56$

Для значення $\beta \approx 0.26$

In [213]:
BSMassSSRel = calculateMassRelation(BS['a1'], BS['a2'], betaSS)
SBMassSSRel = calculateMassRelation(SB['a1'], SB['a2'], betaSS)
MassSSRel = (BSMassSSRel.mean() + 1 / SBMassSSRel.mean()) / 2 

print(MassBBRel)
print(BSMassSSRel.mean())
print(SBMassSSRel.mean())

1.5587535197769675
1.5058609383007333
0.635149381833433


По аналогії з попередніми викладками, було обраховано результати для усіх початкових кутів відхилення та взято середньо-арифметичне по кожному типу експерименту:
- "маленька" куля падає на "велику": $\frac{m_1}{m_2} \approx 0.64$
- "велика" куля падає на "маленьку": $\frac{m_1}{m_2} \approx 1.51$

Середнім арифметичним відношення масс "великої" та  "маленької" кулі є: $\approx 1.56$

Обрахуємо систематичну похибку величини $\beta$
$$\beta = 1 - \frac{1-\cos{\alpha_2}}{1-\cos{\alpha_1}}$$
Знайдемо частинні похідні:
$$\frac{\partial \beta}{\partial \alpha_1} = \frac{(1-\cos{\alpha_2})\sin{\alpha_1}}{(1-\cos{\alpha_1})^2}$$
$$\frac{\partial \beta}{\partial \alpha_2} = -\frac{\sin{\alpha_2}}{1-\cos{\alpha_1}}$$
Підставимо у співвідношення систематичної похибки
$$\Delta \beta = \sqrt{\left(\Delta \alpha_1 \frac{\partial \beta}{\partial \alpha_1}\right)^2 + \left(\Delta \alpha_1 \frac{\partial \beta}{\partial \alpha_2}\right)^2} = \sqrt{\left(\Delta \alpha_1 \frac{(1-\cos{\alpha_2})\sin{\alpha_1}}{(1-\cos{\alpha_1})^2}\right)^2 + \left(- \Delta \alpha_2 \frac{\sin{\alpha_2}}{1-\cos{\alpha_1}}\right)^2}$$
Та відносне відхилення $\delta$. Для $\beta$ - це: $$\delta = \frac{\Delta \beta}{\beta}$$

In [182]:
def systematicBetaError(alpha1, alpha2, deltaAlpha1, deltaAlpha2):
    a1 = alpha1.mean()
    a2 = alpha2.mean()
    da1 = (1 - np.cos(a2)) * np.sin(a1) / np.power((1 - np.cos(a1)),2)
    da2 = np.sin(a2) / (1 - np.cos(a1))
    return np.sqrt(np.power((da1*deltaAlpha1),2) + np.power((da2*deltaAlpha2),2))
    
def relError(delta, values):
    return delta / values.mean()

Для $\Delta \alpha_1 = \Delta \alpha_2 = 0.25^o$ обчислимо систематичну похибку:

In [210]:
experimentsBetaData = pd.concat([BB, SS])
sysBetaError = systematicBetaError(experimentsBetaData['a1'], experimentsBetaData['a2'], np.radians(0.25), np.radians(0.25))
relBBBetaError = relError(sysBetaError, betaBB)
relSSBetaError = relError(sysBetaError, betaSS)
print(sysBetaError)
print(relBBBetaError)
print(relSSBetaError)

0.05731531435402155
0.3751364840891415
0.2220699577282553


Отримаємо: 
- систематична похибка $\Delta \beta  \approx 0.06$
- відносна похибка для $\beta \approx 0.15$: $\delta \approx 0.38 (38\%)$
- відносна похибка для $\beta \approx 0.26$: $\delta \approx 0.22 (22\%)$

Знайдемо систематичну похибку відношення масс(для зручності операцій введемо позначення $\sigma$)
$$\sigma = \frac{m_1}{m_2} = \frac{(1-\beta)(1-\cos{\alpha_1})}{1-\cos{\alpha_2}}$$
Знаходимо частинні похідні
$$\frac{\partial \sigma}{\partial \beta} = \frac{\beta (1-\cos{\alpha_2})}{(1-\beta)^2 (1-\cos{\alpha_1})}$$
$$\frac{\partial \sigma}{\partial \alpha_1} = - \frac{(1-\cos{\alpha_2})\sin{\alpha_1}}{(1-\beta)(1-\cos{\alpha_1})^2}$$
$$\frac{\partial \sigma}{\partial \alpha_2} = \frac{\sin{\alpha_2}}{(1-\beta)(1-\cos{\alpha_1})}$$
Підставляємо у формулу систематичної похибки. Зауважимо, що $\beta$ - це конкрентне значення, яке було знайдено з попередніх експериментів, тому воно не залежить від виміряних у цих роботах значень кутів.
$$\Delta \sigma = \Delta \frac{m_1}{m_2} = \sqrt{\left(\Delta \beta \frac{\partial \sigma}{\partial \beta}  \right)^2 + \left(\Delta \alpha_1 \frac{\partial \sigma}{\partial \alpha_1} \right)^2 + \left(\Delta \alpha_2 \frac{\partial \sigma}{\partial \alpha_2} \right)^2}$$
$$\Delta \sigma = \Delta \frac{m_1}{m_2} = \sqrt{\left(\Delta \beta  \frac{\beta (1-\cos{\alpha_2})}{(1-\beta)^2 (1-\cos{\alpha_1})} \right)^2 + \left(- \Delta \alpha_1 \frac{(1-\cos{\alpha_2})\sin{\alpha_1}}{(1-\beta)(1-\cos{\alpha_1})^2} \right)^2 + \left(\Delta \alpha_2 \frac{\sin{\alpha_2}}{(1-\beta)(1-\cos{\alpha_1})} \right)^2}$$

In [184]:
def systematicMassError(beta, alpha1, alpha2, deltaBeta, deltaAlpha1, deltaAlpha2):
    a1 = alpha1.mean()
    a2 = alpha2.mean()
    dbeta = beta * (1 - np.cos(a2)) / np.power((1 - beta),2) / (1 - np.cos(a1))
    da1 = (1 - np.cos(a2)) * np.sin(a1) / np.power((1 - np.cos(a1)),2) / (1-beta)
    da2 = np.sin(a2) / (1 - np.cos(a1)) / (1-beta)
    return np.sqrt(np.power((dbeta*deltaBeta),2) +  np.power((da1*deltaAlpha1),2) + np.power((da2*deltaAlpha2),2))

Для $\Delta \alpha_1 = \Delta \alpha_2 = 0.25^o$ обчислимо систематичну похибку.
Отримаємо два значення: для $\beta \approx 0.15$ та $\beta \approx 0.26$

In [216]:
experimentsMassData = pd.concat([BS, SB])
sysMassBBError = systematicMassError(betaBB.mean(),experimentsMassData['a1'], experimentsMassData['a2'], sysBetaError ,np.radians(0.25), np.radians(0.25))
sysMassSSError = systematicMassError(betaSS.mean(),experimentsMassData['a1'], experimentsMassData['a2'], sysBetaError ,np.radians(0.25), np.radians(0.25))
relMassBBError = relError(sysMassBBError, MassBBRel)
relMassSSError = relError(sysMassSSError, MassSSRel)
print(sysMassBBError)
print(sysMassSSError)
print(relMassBBError)
print(relMassSSError)

0.07124479864315271
0.08311575932175651
0.04570626320275876
0.05396612610247427


Результати обчислення похибки для різних випадків:
- при $\beta \approx 0.15$ 
     - систематична похибка : $\Delta \frac{m_1}{m_2} \approx 0.07$
     - відносна похибка: $\delta \approx 0.05(5\%)$
- при$\beta \approx 0.26$
    - систематична похибка : $\Delta \frac{m_1}{m_2} \approx 0.08$
    - відносна похибка: $\delta \approx 0.05(5\%)$

## Висновок

Отже, було продемонстровано закон збереження енергії на прикладі куль, що зіштовхуються. Були обраховані значення:
- коефіцієнту розсіяння енергії, що для різних куль з одного матеріалу набув істотно різних значень(для експериментів типу 1: $\beta \approx 0.15 \pm 0.6$,для експериментів типу 2: $\beta \approx 0.26 \pm 0.06$), що можливо пояснити неточністю вимірів.
- співвідношення мас "великої" та "маленької" кулі: 
    - при $\beta \approx 0.15$: $ \frac{m_1}{m_2} = 1.56 \pm 0.07 $
    - при $\beta \approx 0.26$: $ \frac{m_1}{m_2} = 1.56 \pm 0.08 $